app/src/androidTest/kotlin/at/bitfire/davdroid/sync/JtxSyncManagerTest.kt

```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright © All Contributors. See LICENSE and AUTHORS in the root directory for details.
+ */
+
+package at.bitfire.davdroid.sync
+
+import android.content.ContentProviderClient
+import android.content.Context
+import android.content.pm.PackageManager
+import android.os.SystemClock
+import androidx.core.content.ContextCompat
+import androidx.test.platform.app.InstrumentationRegistry
+import at.bitfire.davdroid.db.Collection
+import at.bitfire.davdroid.db.Service
+import at.bitfire.davdroid.network.HttpClient
+import at.bitfire.davdroid.repository.DavServiceRepository
+import at.bitfire.davdroid.resource.LocalJtxCollection
+import at.bitfire.davdroid.resource.LocalJtxCollectionStore
+import at.bitfire.davdroid.sync.account.TestAccount
+import at.bitfire.ical4android.util.MiscUtils.closeCompat
+import at.techbee.jtx.JtxContract
+import dagger.hilt.android.qualifiers.ApplicationContext
+import dagger.hilt.android.testing.HiltAndroidRule
+import dagger.hilt.android.testing.HiltAndroidTest
+import okhttp3.HttpUrl.Companion.toHttpUrl
+import org.junit.After
+import org.junit.Assert.assertEquals
+import org.junit.Assume.assumeTrue
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import java.io.StringReader
+import javax.inject.Inject
+
+
+/**
+ * Ensure you have jtxBoard installed on the emulator, before running these tests
+ */
+@HiltAndroidTest
+class JtxSyncManagerTest {
+
+    @Inject
+    @ApplicationContext
+    lateinit var context: Context
+
+    @Inject
+    lateinit var httpClientBuilder: HttpClient.Builder
+
+    @Inject
+    lateinit var serviceRepository: DavServiceRepository
+
+    @Inject
+    lateinit var localJtxCollectionStore: LocalJtxCollectionStore
+
+    @Inject
+    lateinit var jtxSyncManagerFactory: JtxSyncManager.Factory
+
+    @get:Rule
+    val hiltRule = HiltAndroidRule(this)
+
+    private val account = TestAccount.create()
+
+    private lateinit var provider: ContentProviderClient
+    private lateinit var syncManager: JtxSyncManager
+    private lateinit var localJtxCollection: LocalJtxCollection
+
+    @Before
+    fun setUp() {
+        hiltRule.inject()
+
+        val context = InstrumentationRegistry.getInstrumentation().targetContext
+
+        // Check jtxBoard is installed; skip test otherwise
+        assumeTrue("jtxBoard not installed", isJtxBoardInstalled())
+
+        // Grant content provider permissions at runtime; skip tests on failure
+        val permissions = listOf("at.techbee.jtx.permission.READ", "at.techbee.jtx.permission.WRITE")
+        val permissionsGranted = tryGrantPermissions(context.packageName, permissions)
+        assumeTrue("Missing permissions", permissionsGranted)
```

Comment:
Can we instead use `GrantPermissionRule`? I have seen that it throws a `SecurityException` when jtx Board is not installed and it thus can't grant the jtx Board permissions.

However I think it should be possible to wrap it in a custom `IgnoreExceptionRule` or something like that.

Reply:
That would be very nice. I have currently no idea how that would work though. Where can I find documentation on custom rules?

Reply:
- https://www.baeldung.com/junit-4-rules
- https://github.com/junit-team/junit4/wiki/Rules#custom-rules

There's a bit information about nesting rules. It's a bit cumbersome but I think it should be doable.

Reply:
Alright. At first I thought chaining an "ignore exceptions" rule after the grant permissions rule, should work too, but it does not. Wrapping the rule in another one does work though :+1: